File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/immutability/java/ImmutabilityExample.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.immutability.java;

import pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.readability.java.User;

import java.util.List;

public class ImmutabilityExample {

    public User createUser() {
        // final prevents reassignment of the variable.
        final String firstName = "John";

        // Regular Java variables are mutable by default.
        int age = 25;
        age += 1;

        // List.of creates an immutable list.
        List<String> roles = List.of("USER", "ADMIN");

        return new User(
                firstName,
                "Doe",
                age,
                roles
        );
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/immutability/kotlin/ImmutabilityExample.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.immutability.kotlin

import pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.readability.kotlin.User

class ImmutabilityExample {

    fun createUser(): User {
        // val creates a read-only reference.
        // It cannot be reassigned after initialization.
        val firstName = "John"

        // Mutable reference.
        // It can be reassigned.
        var age = 25
        age += 1

        // listOf creates a read-only list interface.
        val roles = listOf("USER", "ADMIN")

        return User(
            firstName = firstName,
            lastName = "Doe",
            age = age,
            roles = roles
        )
    }
}